sqlmesh/core/signal.py

```diff
@@ -33,3 +38,42 @@ class signal(registry_decorator):
 
 
 SignalRegistry = UniqueKeyDict[str, signal]
+
+
+@signal()
+def freshness(batch: DatetimeRanges, snapshot: Snapshot, context: ExecutionContext) -> bool:
+    if context.is_restatement_plan:
+        return True
+
+    deployability_index = context.deployability_index
+    adapter = context.engine_adapter
+
+    if not deployability_index or not adapter.SUPPORTS_METADATA_TABLE_LAST_MODIFIED_TS:
+        return True
+
+    last_altered_ts = (
+        snapshot.last_altered_ts
+        if deployability_index.is_deployable(snapshot)
+        else snapshot.dev_last_altered_ts
+    )
+    if not last_altered_ts:
+        return True
+
+    parent_snapshots = {context.snapshots[p.name] for p in snapshot.parents}
+    if len(parent_snapshots) != len(snapshot.node.depends_on) or not all(
+        p.is_external for p in parent_snapshots
+    ):
+        # The mismatch can happen if e.g an external model is not registered in the project
+        return True
+
+    # Finding new data means that the upstream depedencies have been altered
+    # since the last time the model was evaluated
+    upstream_dep_has_new_data = any(
+        upstream_last_altered_ts > last_altered_ts
+        for upstream_last_altered_ts in adapter.get_table_last_modified_ts(
+            [p.name for p in parent_snapshots]
```

Comment:
This looks wrong:
1. Why do we check tables of non-external model snapshots? Shouldn't we rely on their `last_altered_ts` instead?
2. Why are we not using the physical table name?

Reply:
Ok, I see. Thanks